Add format validation for invoice attributes

mikaellisselgard opened the pull request:
PayEx have validations for the invoice attributes, we reflect those
validations here to avoid sending invalidig attributes to PayEx. This
patch also introduces a resolve option, where you can pass a Proc to try
to format the attribute correctly and retry the validation. For example
shorten the customer name if it's too long. See below for validations:

http://www.payexpim.com/technical-reference/pxorder/purchaseinvoiceprivate/

mikaellisselgard committed:
Add format validation for invoice attributes

PayEx have validations for the invoice attributes, we reflect those
validations here to avoid sending invalidig attributes to PayEx. This
patch also introduces a resolve option, where you can pass a Proc to try
to format the attribute correctly and retry the validation. For example
shorten the customer name if it's too long. See below for validations:

http://www.payexpim.com/technical-reference/pxorder/purchaseinvoiceprivate/

hovsater commented on lib/payex/api/pxorder.rb:
```diff
@@ -103,65 +103,83 @@ def PurchaseInvoicePrivate(params)
       'socialSecurityNumber' => {
-        signed: true
+        signed: true,
+        format: /^[a-zA-Z0-9_:!;\\”#<>=?\\[\\]@{}´\n\r %-\/À-ÖØ-öø-úü]{,15}$?/
```
I'm not entirely sure this regex is correct. Try this instead `\A[a-zA-Z0-9_:!;\"#<>=?\\[\\]@{}´\n\r %-\/À-ÖØ-öø-úü]*\z`.

mikaellisselgard committed:
Add sample product code

When validations matching PayEx validations was introduced it caused a
validation error when using the product number as product code.

hovsater commented on lib/payex/api.rb:
```diff
@@ -110,4 +112,16 @@ def parse_transaction_status(status)
+  end
+
+  def parse_with_resolve!(param, options)
+    resolve = options[:resolve]
```
Could also be written as:

```ruby
resolve = options.fetch(:resolve) { param_error! 'resolve parameter required' }
```

hovsater commented on lib/payex/api.rb:
```diff
@@ -110,4 +112,16 @@ def parse_transaction_status(status)
+  def parse_with_resolve!(param, options)
+    resolve = options[:resolve]
+    param_error! %{resolve parameter required} if resolve.nil?
+    result = resolve.call(param) if resolve.is_a? Proc
```
I think asking if it responds to `#call` is better. What do you think?

hovsater commented:
Could also be written as:

```ruby
def parse_param(param, options)
  value = param || options.fetch(:default) { param_error! 'parameter required' }
  value = value.call if value.respond_to?(:call)
  
  return value if options[:format] == value
  
  if is_resolver?(options[:resolve])
    parse_param(options[:resolver].call(value), options.except(:default, :resolve))
  else
    param_error! %{must match #{options[:format].inspect}}
  end
end


def is_resolver?(resolver)
  resolver.respond_to?(:call)
end
```

hovsater commented on lib/payex/api/pxorder.rb:
```diff
@@ -103,65 +103,83 @@ def PurchaseInvoicePrivate(params)
-        signed: true
+        signed: true,
+        format: /\A[a-zA-Z0-9_:!;\"#<>=?\\[\\]@{}´\n\r %-\/À-ÖØ-öø-úü]{,35}\z/,
+        resolve: proc { |v| v[0..34].strip }
```
could also be written as `-> (v) { }`

hovsater commented on spec/api_spec.rb:
```diff
@@ -57,4 +57,115 @@
+        it 'calls the proc' do
+          options = {
+            format: /^[a-z]{,3}$/,
+            resolve: proc { |v| v[0..2].strip }
```
Extract the proc to a variable, and the use it in the options. It reads better.

```ruby
resolver = -> (v) { v[0..2].strip }

options = { format: /^[a-z]{,3}$/, resolve: resolver }

resolver.should_receive(:call).with('too long param').and_return('too')
```

mikaellisselgard commented:
@KevinSjoberg Vettiga förslag, fixat här: https://github.com/standout/payex-ruby/pull/4/commits/88c99e5fc0cc7b9d57d8279781872097d0e46ac2

Istanful commented on spec/api_spec.rb:
```diff
@@ -57,4 +57,85 @@
+  describe '#parse_param' do
+    context 'when format validation fails' do
+      context 'when resolve proc is present' do
+        it 'calls the proc' do
```
Why shoulda matchers?

Provided you can I would probably write something like:
```ruby
resolver = -> (v) { v[0..2].strip }
options = { format: /^[a-z]{,3}$/, resolve: resolver }
allow(resolver).to receive(:call).and_return('too')

PayEx::API.parse_param('too long param', options)

expect(resolver).to have_received(:call).with('too long param')
```

mikaellisselgard replied:
It's because we're running rspec 2.11.

Istanful replied:
Gotcha. I suspected something like that. Feel free to merge :)

Istanful reviewed (approved):
Fix if possible and you want. Good otherwise 👍
